// @tact-lang/opcode/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTextPrinter = exports.opcodeToString = exports.loadOpcode = exports.isOpCodeWithArgs = exports.decompileAll = exports.decompile = void 0;
var decompiler_1 = require("./decompiler/decompiler");
Object.defineProperty(exports, "decompile", { enumerable: true, get: function () { return decompiler_1.decompile; } });
var decompileAll_1 = require("./decompiler/decompileAll");
Object.defineProperty(exports, "decompileAll", { enumerable: true, get: function () { return decompileAll_1.decompileAll; } });
var opcodes_gen_1 = require("./codepage/opcodes.gen");
Object.defineProperty(exports, "isOpCodeWithArgs", { enumerable: true, get: function () { return opcodes_gen_1.isOpCodeWithArgs; } });
var loadOpcode_1 = require("./codepage/loadOpcode");
Object.defineProperty(exports, "loadOpcode", { enumerable: true, get: function () { return loadOpcode_1.loadOpcode; } });
var opcodeToString_1 = require("./codepage/opcodeToString");
Object.defineProperty(exports, "opcodeToString", { enumerable: true, get: function () { return opcodeToString_1.opcodeToString; } });
var printer_1 = require("./decompiler/printer");
Object.defineProperty(exports, "createTextPrinter", { enumerable: true, get: function () { return printer_1.createTextPrinter; } });
