// @ton-community/blueprint/dist/compile/compile.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compile = exports.doCompile = void 0;
const func_js_1 = require("@ton-community/func-js");
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
const ton_core_1 = require("ton-core");
const paths_1 = require("../paths");
const compiler_1 = require("@tact-lang/compiler");
const OverwritableVirtualFileSystem_1 = require("./OverwritableVirtualFileSystem");
async function getCompilerConfigForContract(name) {
    var _a;
    require('ts-node/register');
    const mod = await (_a = path_1.default.join(paths_1.WRAPPERS_DIR, name + '.compile.ts'), Promise.resolve().then(() => __importStar(require(_a))));
    if (typeof mod.compile !== 'object') {
        throw new Error(`Object 'compile' is missing`);
    }
    return mod.compile;
}
async function doCompileFunc(config) {
    const cr = await (0, func_js_1.compileFunc)(config);
    if (cr.status === 'error')
        throw new Error(cr.message);
    return {
        lang: 'func',
        code: ton_core_1.Cell.fromBase64(cr.codeBoc),
    };
}
function findTactBoc(fs) {
    let buf = undefined;
    for (const [k, v] of fs) {
        if (k.endsWith('.code.boc')) {
            buf = v;
            break;
        }
    }
    if (buf === undefined) {
        throw new Error('Could not find boc in tact compilation result');
    }
    return ton_core_1.Cell.fromBoc(buf)[0];
}
async function doCompileTact(config, name) {
    const fs = new OverwritableVirtualFileSystem_1.OverwritableVirtualFileSystem(process.cwd());
    const res = await (0, compiler_1.build)({
        config: {
            name: 'tact',
            path: config.target,
            output: path_1.default.join(paths_1.BUILD_DIR, name),
            options: config.options,
        },
        stdlib: '/stdlib',
        project: fs,
    });
    if (!res) {
        throw new Error('Could not compile tact');
    }
    const code = findTactBoc(fs.overwrites);
    return {
        lang: 'tact',
        fs: fs.overwrites,
        code,
    };
}
async function doCompileInner(name, config) {
    if (config.lang === 'tact') {
        return await doCompileTact(config, name);
    }
    return await doCompileFunc({
        targets: config.targets,
        sources: config.sources ?? ((path) => (0, fs_1.readFileSync)(path).toString()),
        optLevel: config.optLevel,
    });
}
async function doCompile(name) {
    const config = await getCompilerConfigForContract(name);
    if (config.preCompileHook !== undefined) {
        await config.preCompileHook();
    }
    const res = await doCompileInner(name, config);
    if (config.postCompileHook !== undefined) {
        await config.postCompileHook(res.code);
    }
    return res;
}
exports.doCompile = doCompile;
async function compile(name) {
    const result = await doCompile(name);
    return result.code;
}
exports.compile = compile;
