// @tact-lang/compiler/dist/config/features.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.featureEnable = exports.featureEnabled = exports.enabledExternals = exports.enabledMaterchain = exports.enabledDebug = exports.enabledInline = void 0;
const context_1 = require("../context");
const featureStore = (0, context_1.createContextStore)();
function enabledInline(ctx) {
    return featureEnabled(ctx, 'inline');
}
exports.enabledInline = enabledInline;
function enabledDebug(ctx) {
    return featureEnabled(ctx, 'debug');
}
exports.enabledDebug = enabledDebug;
function enabledMaterchain(ctx) {
    return featureEnabled(ctx, 'masterchain');
}
exports.enabledMaterchain = enabledMaterchain;
function enabledExternals(ctx) {
    return featureEnabled(ctx, 'external');
}
exports.enabledExternals = enabledExternals;
function featureEnabled(ctx, key) {
    return featureStore.get(ctx, key) === true;
}
exports.featureEnabled = featureEnabled;
function featureEnable(ctx, key) {
    return featureStore.set(ctx, key, true);
}
exports.featureEnable = featureEnable;
